driver-core/src/test/functional/com/mongodb/client/syncadapter/SupplyingCallback.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.client.syncadapter;
+
+import com.mongodb.annotations.ThreadSafe;
+import com.mongodb.internal.async.SingleResultCallback;
+import com.mongodb.lang.Nullable;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+@ThreadSafe
+public final class SupplyingCallback<R> implements SingleResultCallback<R>, Supplier<R> {
```

Comment:
Initially this class was added in the successive PR https://github.com/mongodb/mongo-java-driver/pull/784, but I needed it here to write unit tests for `LoopState`/`RetryState`. So I had to "back port" the class here, and will pull the changes to https://github.com/mongodb/mongo-java-driver/pull/784 to resolve conflicts.